### PR TITLE
Add support for proto2 extensions

### PR DESCRIPTION
**Pull request by EPronovost**
Adds support for message extensions.  Now if there is a message 
```
message Bar {
  extend Foo { optional Bar ext = 2020; }
  
  optional bool flag = 1;
}
```
then `Bar.ext` will no longer raise a mypy error.  Furthermore, we are able to correctly type extension messages, so for something like
```
foo = Foo()
bar = foo.Extensions[Bar.ext]
```
the inferred type for `bar` is `Bar`.

### CODE REVIEW

File: test/test_generated_mypy.py
Comment (nipunn1313) [inline test/test_generated_mypy.py:327]: May want to consider mypy-typing the `Extensions` variable on all messages to be a
```
Mapping[FieldDescriptor, google.protobuf.Message]
```

I'd also appreciate some kind of `assert isinstance(Extensions1.ext, FieldDescriptor)` to confirm that we're mypying it  to the correct type.

File: test/test_generated_mypy.py
Comment (nipunn1313) [inline test/test_generated_mypy.py:330]: If `Extensions` is typed, then e2 can by typed as a Message here.
It'll probably require some manual casting by the caller to get the appropriate message type, but I don't really see a way around that. I think it's probably better to force the caller to do an explicit type cast from Message to the appropriate Message type, over silently returning Any here.

In any case, what you have so far is better than what we had before, so I'm ok to run with it.

**Review by nipunn1313 — CHANGES_REQUESTED** [review]
This looks great! Couple of comments around the testing, then we can go with it.

Trying to think of how we can add a negative test here - to confirm that this mypy typing catches bugs - has better typing than Any. Perhaps in `test_negative.py` we can add something of the sort
```
x = 5
x = Bar.ext   # E
```

Where we'd expect a type error of a FieldDescriptor being assigned to an int.

**Comment by nipunn1313** [comment]
Thank you so much for the contribution!
Must mention that Dropbox requires that you agree to the CLA before it can be accepted https://opensource.dropbox.com/cla/

Thank you!

**Comment by nipunn1313** [comment]
Would also ask that you update CHANGELOG.md with the new change in the "Upcoming" section

File: test/test_generated_mypy.py
Comment (EPronovost) [reply test/test_generated_mypy.py]: I figured out a way to improve type inference for extensions.  Now we type `e1` as `Extensions1` explicitly.  Doing so required overwriting the `Extensions` defined for `Message`s in the typeshed.
I also added such `assert` statements.

File: test/test_generated_mypy.py
Comment (EPronovost) [reply test/test_generated_mypy.py]: I figured out a way around it using generics :).  Now `e2` is inferred to be an `Extensions2`.

**Comment by EPronovost** [comment]
The PR got more complicated, but we now have correct type inference for extensions.

**Comment by nipunn1313** [comment]
Hmm - looks like this diff has developed some additional complexity for typing the extensions dict - and potentially has some points up for debate. If you were to split the PR into uncontroversial part (w/o typing extensions dict) - we can move forward with that easily, so we can focus discussion on this more controversial complexity and find the right way to solve it.

I don't want to hold up this entire diff on this discussion - since it's clearly adding value!
Thank you so much for doing this work!

**Comment by EPronovost** [comment]
@nipunn1313 I made a more minimal PR here: https://github.com/dropbox/mypy-protobuf/pull/149
What would you like to discuss about this PR?

**Review by nipunn1313 — COMMENTED** [review]
Sorry Forgot to hit submit on the comments! Here they are.

File: python/mypy_protobuf.py
Comment (EPronovost) [reply python/mypy_protobuf.py]: Yeah I removed this in favor of a template file.

File: python/mypy_protobuf.py
Comment (EPronovost) [reply python/mypy_protobuf.py]: Previously the `_ExtensionDict` we defined did not inherit from the `ExtensionDict` defined in the standard typeshed, so although the methods were similar they were different objects.  I've changed this to fix it (note that there are still override ignores for the methods we're overwriting).

File: generated/_typeshed_mypy_protobuf.pyi.expected
Comment (EPronovost) [reply generated/_typeshed_mypy_protobuf.pyi.expected]: > 1 - Why does this need to be autogenerated? Seems like it potentially should belong either directly in typeshed itself, or inside the mypy_protobuf library as something to be imported

Yes I agree the autogeneration is not the best solution.  I can think of 3 possibilities:

* Generate a file as part of the protoc plugin (i.e. what we have currently).  I changed it to use a template file instead of having python code to write the file on the fly.  The benefit of this is that we can ensure these stubs will be present whenever `mypy` is run and don't add any new package requirements.

* As part of the `mypy_protobuf` library.  This simplifies the generated code and still ensures that the same version of the base classes used to generate stubs is used to evaluate them (assuming the user doesn't change library versions in between).  However, I think it also means that `mypy_protobuf` would need to be present whenever `mypy` is run, which is not currently a requirement.

* As part of the `typeshed`.  This addresses the issue of needing `mypy_protobuf` present when running `mypy`.  However, it means that we'd require the new `mypy` version to have these stubs be present.  It would also pose difficulties if we want to change anything about this file in the future.

There may be a way to do either of the latter two options in a "soft" way; i.e. fallback to the minimal extension support already in `master` if the respective dependency is not present.  I don't have a ton of experience publishing to the `typeshed` to know if/how they support API changes across package versions.

> 2 - Can you explain what value this is adding over a Dict[ContainerMessage, ExtenderMessage]

Two reasons.  `Message.Extensions` is not a dictionary, and does not support the various other methods associated with one (e.g. you can't do `foo.Extensions.values()`).  Additionally, I don't think using a static typing would allow for the polymorphism we currently have (i.e. `_ExtensionDict` is a class parametrized by `_ContainerMessageT` and `__{gs}estitem__` is a method of that class parametrized by `_ExtenderMessageT`).  The actual source code is [here](https://github.com/protocolbuffers/protobuf/blob/e667bf6eaaa2fb1ba2987c6538df81f88500d030/python/google/protobuf/internal/extension_dict.py#L63).  I'm not sure why the typeshed implementation does not include the other methods; I added all the public ones here.

File: python/setup.py
Comment (EPronovost) [inline python/setup.py:11]: I don't have any experience making a PyPI package... if there's a better way to do this let me know.

File: generated/_typeshed_mypy_protobuf.pyi.expected
Comment (nipunn1313) [reply generated/_typeshed_mypy_protobuf.pyi.expected]: Ah! We can totally add these stubs to typeshed - as long as the stubs match the source code.
We have lots of precedent doing that (much of the google/protobuf typeshed stubs were contributed by dropboxers / folks who work on mypy-protobuf)

Source: https://github.com/protocolbuffers/protobuf/blob/e667bf6eaaa2fb1ba2987c6538df81f88500d030/python/google/protobuf/internal/extension_dict.py#L63

Typeshed: https://github.com/python/typeshed/tree/master/third_party/2and3/google/protobuf/internal
https://github.com/python/typeshed/tree/master/third_party/2and3/google/protobuf/internal

You can also test your typeshed change + mypy-protobuf change within mypy-protobuf by doing
```
CUSTOM_TYPESHED_DIR=~/path/to/your/typeshed/fork ./run_test.sh
```

Though one gotcha - when developing you should run the typeshed tests first - because errors within typeshed get suppressed when you use CUSTOM_TYPESHED_DIR.

It does mean that we'll have to wait for another release of mypy before we launch the next version of mypy-protobuf, but I think that's ok! #140 is similarly waiting on a typeshed change (you can look at that as an example)

File: python/mypy_protobuf.py
Comment (nipunn1313) [reply python/mypy_protobuf.py]: Seems like we may be able to avoid this by updating the standard typeshed to be more like what we want!

**Comment by nipunn1313** [comment]
Here are some examples of updates to typeshed for sake of mypy-protobuf
https://github.com/python/typeshed/pull/4265/files
https://github.com/python/typeshed/pull/3996/files
https://github.com/python/typeshed/pull/3995/files

File: generated/_typeshed_mypy_protobuf.pyi.expected
Comment (EPronovost) [reply generated/_typeshed_mypy_protobuf.pyi.expected]: Interesting... I didn't realize the typeshed stubs for protobuf came from this same group (thank you!).  Are there any further comments you have on the current code before moving `typeshed.pyi` to the official typeshed?

File: python/setup.py
Comment (EPronovost) [reply python/setup.py:11]: Will revert back once the typeshed is updated.

**Comment by EPronovost** [comment]
This will be blocked on https://github.com/python/typeshed/pull/4619

**Comment by nipunn1313** [comment]
the typeshed change landed, and I landed a mypy->typeshed pin bump afterward!

I spoke with the mypy maintainers (they're also at Dropbox), and they mentioned that it may be until around November before there's a mypy release w/ your change.
Feel free to update https://github.com/dropbox/mypy-protobuf/blob/master/run_test.sh#L36 to a git revision to unblock your change. In November (or whenever), I'll change it back before releasing mypy-protobuf

From
```
python3 -m pip install git+https://github.com/python/mypy.git@v0.780
```
to
```
python3 -m pip install git+https://github.com/python/mypy.git@7273e9ab1664b59a74d9bd1d2361bbeb9864b7ab
```
referencing https://github.com/python/mypy/commit/7273e9ab1664b59a74d9bd1d2361bbeb9864b7ab

File: test/test_generated_mypy.py
Comment (nipunn1313) [inline test/test_generated_mypy.py:358]: This comment might leave some confusion as to the why, I think more specifically, explaining what "x" should be typed as will help understand why "y" is an `Any` here.

Specifically, to add clarity that `y` is `Any` because `x` came out of an iterator, which can't know the value type.

**Comment by nipunn1313** [comment]
This is awesome!!! Thank you so much for the contributions here.

I left one message about a comment in the tests. Going to go ahead and merge it - feel free to add another PR to update the comment if you're excited.

**Pull request merged**
